src/engine/TPA.cc

```diff
@@ -982,16 +983,101 @@ bool TPASplit::verifyExactPower(unsigned short power) const {
+    //   leftInvariants /\ transition /\ getNextVersion(currentLevelTransition) =>
+    //     shiftOnlyNextVars(currentLevelTransition);
+    while (candidates.size() > 128) {
+        for (int i = candidates.size() - 1; i >= 1; i -= 2) {
+            PTRef n_f = logic.mkAnd(candidates[i], candidates[i - 1]);
+            candidates.pop();
+            candidates.pop();
+            candidates.push(n_f);
+            if (candidates.size() <= 128) { break; }
+        }
+    }
+    //    invCandidates.append(conjuncts);
+    //    Atr(x, x') /\ tr(x', x'') => Atr(x, x'')
+    //    or
+    //    Atr(x', x'') /\  tr(x, x') => Atr(x, x'')
+    //    Push transition once and for all
+    //    While loop externally, because we may drop smth important
+    PTRef goal = shiftOnlyNextVars(logic.mkAnd(candidates));
+
+    if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+        solver.insertFormula(logic.mkAnd({logic.mkAnd(candidates), logic.mkNot(goal)}));
+    } else {
+        if (allignment == SafetyExplanation::FixedPointType::LEFT) {
+            solver.insertFormula(logic.mkAnd({getNextVersion(logic.mkAnd(candidates)), logic.mkNot(goal)}));
+        }
+    }
+    while (solver.check() == s_True) {
+        for (int i = candidates.size() - 1; i >= 0; i--) {
+            PTRef cand = candidates[i];
+            solver.pop();
+            solver.push();
+            if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+                solver.insertFormula(logic.mkAnd({logic.mkAnd(candidates), logic.mkNot(shiftOnlyNextVars(cand))}));
+            } else {
+                if (allignment == SafetyExplanation::FixedPointType::LEFT) {
+                    solver.insertFormula(
+                        logic.mkAnd({getNextVersion(logic.mkAnd(candidates)), logic.mkNot(shiftOnlyNextVars(cand))}));
+                }
+            }
+            if (solver.check() == s_True) {
+                Model model = *solver.getModel();
+                candidates[i] = candidates[candidates.size() - 1];
+                candidates.pop();
+            }
+        }
+        solver.pop();
+        solver.push();
+        goal = shiftOnlyNextVars(logic.mkAnd(candidates));
+        if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+            solver.insertFormula(
+                logic.mkAnd({logic.mkAnd(candidates), logic.mkNot(goal)}));
+        } else {
+            solver.insertFormula(logic.mkAnd({getNextVersion(logic.mkAnd(candidates)), logic.mkNot(goal)}));
+        }
+    }
+    for (auto cand : candidates) {
+        if (allignment == SafetyExplanation::FixedPointType::RIGHT) {
+            if (std::find(rightInvariants.begin(), rightInvariants.end(), cand) != rightInvariants.end()) { continue; }
+            rightInvariants.push(cand);
+        } else {
+            if (std::find(leftInvariants.begin(), leftInvariants.end(), cand) != leftInvariants.end()) { continue; }
+            leftInvariants.push(cand);
+        }
+    }
```

Comment:
This should be better to do at the beginning, then you don't have to do SMT queries for this candidates at all.
Also, it would be more efficient to keep the invariants in a set as well as in a vector. So that the membership check would be much cheaper.